### DISCUSSION TITLE
Please Note: SysEx file numbering has changed in the latest builds!

### DISCUSSION THREAD

**Discussion by diyelectromusic** [probonopd/MiniDexed#482] [General]
There has been a fair bit of work on SysEx voice/bank file loading recently, as you might have spotted if you've been following the discussion here: #456, issues #391 #455 #457 #458 #459 #460 #468 and changes here: #461 #464 #466 #473 (I think that is all of them!).

New functionality includes:
* Faster loading of SysEx files by supporting the use of subdirectories (recommend <100 files per directory for fast loading).
* Up to 16384 banks now supported.
* MIDI BANKSEL MSB/LSB now supported to select any of the 16384 banks.
* Empty banks and voices are skipped in the menus,
* "Headerless" SysEx files are now supported too (although disabled by default).
* MIDI Program Change messages can now select up to 128 voices across four consecutive banks if required.

And on top of this there is a huge effort going on to curate a library of known useful voices - watch this space!

Most of the changes are behind the scenes, but one that affects backwards compatibility relates to the numbering of SysEx files.

**In order to make sure the number shown on the menu and the number in the filename are the same, yet still support the MIDI standard of 1-indexed numbering for people, but 0-indexed numbering for the protocol, _SysEx file number 000000 is no longer supported_.**

This means that if you have existing SysEx files that have been numbered for you, they will need renumbering starting from 000001 if you want to keep them in the same order.  Of course if you don't need them in the same order you can rename 000000 to any other free number between 1 and 16384.

The wiki has been updated, but the provided script to download and name SysEx files currently still starts from 0.  It will need to change to start from 1, but there are other changes happening to that script atm, so watch this space on that one too...

But I just wanted to get the renumbering issue "out there" in case anyone updates MiniDexed and wonders why things have changed!

Kevin

**Comment by probonopd** [comment]
I think this slipped through when I merged the PR. Why don't we want MIDI voice 0 in file 0?

**Reply by diyelectromusic** [reply]
Because people would name a file 00002_MyBank.syx but it would show up in the menus as 3 MyBank.  The initial thought was just to make them the same, but the issue is that MIDI is quite explicit in the specification that banks should be presented to the user as 1-indexed, but the protocol should send the numbers as 0-indexed.  This means that to make the "user view" consistent and still compliant with MIDI (and therefore any external sysEx/voice editor tool/etc) we either keep the filenames out of sync with the menu, or stop people using bank number 0.

Some discussion here: https://github.com/probonopd/MiniDexed/discussions/456#discussioncomment-5542924

Kevin

**Reply by probonopd** [reply]
Do you know there any "real" instruments that use numbered files? How are they numbering them?

**Reply by diyelectromusic** [reply]
Real instruments have banks and if numbered they will start at 1.  Only computer people start counting things at 0 ;)

But the MIDI spec is quite explicit. And either way it is very confusing for a user if their file has one number, but the menus and any MIDI SysEx tools refer to it as a different number.  Making them consistent is I think really quite important.

Kevin

**Reply by probonopd** [reply]
> Making them consistent is I think really quite important.

I agree. Only I was thinking that it is confusing if the files and the MIDI messages use different numbers (MIDI and files being the "technical" layer, while menus are the "user visible" layer).
Well, can't get all 3 consistent. Can only get 2 consistent. If you think that what we have now is more in line with the MIDI spec with what we had before, then let's use the new scheme.

**Reply by miotislucifugis** [reply]
sorry if this is a naive question, but is it possible to do away with the file name numbering altogether, and just have a number assigned based on its position on the card; alphabetical, date added, etc?

**Reply by diyelectromusic** [reply]
Most people don't ever see the "on the wire" MIDI message.  It is the same with Program Numbers.  The MIDI spec states the user should be presented with values 1..128, but "on the wire" they are 0..127. It is very clear on this topic.  For bank numbers the same applies but for 0..16383 shown as 1..16384 - see https://github.com/probonopd/MiniDexed/discussions/456#discussioncomment-5543531

The files are really a user-layer, in that the user will store them on the SD card themselves and expect to be able to find them again in the menus and in any external SysEx librarian type tool they might be using externally via MIDI, which will use 1-indexed numbering too.

Kevin

**Reply by diyelectromusic** [reply]
> sorry if this is a naive question, but is it possible to do away with the file name numbering altogether, and just have a number assigned based on its position on the card; alphabetical, date added, etc?

The ordering of files in a directory is a bit of a black art.  It would have to be a more explicit criteria such as alphabetical, ASCII order or some other "obvious" sorting criteria... (did you ever try using a file-based MP3 player before they started to recognise meta data?  It was a nightmare!!).

I suspect people would probably try to add numbered prefixes in order to get some kind of order anyway... so we may as well keep it an explicit relationship rather than an implied one I think.

Kevin

**Reply by miotislucifugis** [reply]
I see.   Can it be auto generated?   There is something nice about just being able to dump some syx files onto a card and trying them out without having to rename everything.    (apologies if i sound like a lazy ingrate ;) you guys have been doing great work! )

**Reply by probonopd** [reply]
Should be easy to script this. Can you run Python or .sh scripts on your machine?

**Reply by diyelectromusic** [reply]
I know its not the same, but it will also accept single voices over MIDI from a SysEx tool like MidiOx, or even Dexed running on a PC... although there isn't a way to save them at the moment.

**Reply by miotislucifugis** [reply]
"(did you ever try using a file-based MP3 player before they started to recognise meta data? It was a nightmare!!)."

ha yes. i remember... also SD cards.  Fatsort  works pretty well for reordering them.       just saying it would be cool do dump one of the massive libraries of patches on without curating them one by one...i think most people would just remember the number for the bank they liked, rather than wanting it to be saved as a specific number, if that makes sense.

**Reply by diyelectromusic** [reply]
It might be possible to have this as an option if it is something people would like I guess...  It would be a relatively minor change, but I don't think both schemes should be active at once :)

The only issue I can see is that we now recommend using subdirectories if there are more than (say) 100 files otherwise they are very slow to load, so the ordering of directories would have to be taken into account too.  And I think it traverses them "depth first" (trying to remember how I did it!?) so adding new directories could significantly reorder things!

Kevin

**Reply by miotislucifugis** [reply]
> Should be easy to script this. Can you run Python or .sh scripts on your machine?

yeah sure... just thinking about general usability and how it works on other instruments.  been playing with a headless M8 setup recently... Samples, instruments, songs, etc can all just be put on the card as-is and when you look at the contents of a file you see the files in whatever order they are in on the card, usually in order they were added.  its not that big of a deal and one can reorder the card contents if they like.  I realize that minidexed is a different thing and officially DIY, but I can see how being required to rename and organize the files might be a bit of cramp in the work flow.

**Reply by miotislucifugis** [reply]
> It might be possible to have this as an option if it is something people would like I guess... It would be a relatively minor change, but I don't think both schemes should be active at once :)
> 
> The only issue I can see is that we now recommend using subdirectories if there are more than (say) 100 files otherwise they are very slow to load, so the ordering of directories would have to be taken into account too. And I think it traverses them "depth first" (trying to remember how I did it!?) so adding new directories could significantly reorder things!
> 
> Kevin

one thing at a time.   I dont think it a serious issue.. just tossing it out there
